src/include/sof/measure.h

```diff
@@ -0,0 +1,199 @@
+ */
+#define PERF_REPORT_FMT " - avg perf: %u us, %u cycles"
+
+/**
+ * Statistics types
+ */
+enum measure_stats {
+	MEASURE_STAT_MIN = 0,
+	MEASURE_STAT_AVG,
+	MEASURE_STAT_MAX,
+
+	MEASURE_NUM_STATS
+};
+
+/**
+ * Measurement sample
+ */
+struct measure_sample {
+	uint64_t time_delta;
+	uint32_t ccount_delta;
+};
+
+/**
+ * Measurement context
+ */
+struct measure_context {
+	uint32_t steps;
+	uint32_t avg_shift;
+	uint32_t cur_step;
+	uint64_t begin_time;
+	uint64_t begin_ccount;
+	uint64_t last_report;
+	struct measure_sample *samples;
+	struct measure_sample stats[MEASURE_NUM_STATS];
+};
+
+/**
+ * \brief Initializes measurement and returns context
+ *
+ * The context is used to store current statistics and measurement status,
+ * therefore it should be kept and reused with the following calls.
+ *
+ * \param[in] steps_log2 History length for the computation of running average,
+ *                       in the form of a power of 2: log2(steps)
+ * \return Measurement context
+ */
+struct measure_context measure_init(uint32_t steps_log2);
+
+/**
+ * \brief Invalidates and frees context
+ * \param[in] context Measurement context
+ */
+void measure_free(struct measure_context *context);
+
+/**
+ * \brief Sets the measurement reference
+ *
+ * Timing mesurements will be taken in relation to the point of calling
+ * this function.
+ *
+ * \param[in] context Measurement context
+ */
+void measure_begin(struct measure_context *context);
+
+/**
+ * \brief Takes the measurement
+ *
+ * Performs timing measurements in relation to previously set reference
+ * (see measure_begin()) and recalculates statistics.
+ *
+ * \param[in] context Measurement context
+ */
+void measure_end(struct measure_context *context);
+
+/**
+ * \brief Returns microseconds of a given statistic type
+ * \param[in] context Measurement context
+ * \param[in] stat_type Statistic type (min, avg, max)
+ * \return Time in microseconds
+ */
+uint64_t measure_get_microseconds(struct measure_context *context,
+				  int stat_type);
+
+/**
+ * \brief Returns ccount of a given statistic type
+ * \param[in] context Measurement context
+ * \param[in] stat_type Statistic type (min, avg, max)
+ * \return Number of cycles
+ */
+uint64_t measure_get_ccount(struct measure_context *context, int stat_type);
+
+/**
+ * \brief Reports the measurement
+ *
+ * Will report the measurements every interval milliseconds
+ * in the form of a trace containing average timing:
+ *
+ * "<prefix> - avg perf: %u us, %u cycles"
+ * The prefix may contain up to 2 32-bit integer parameters.
+ *
+ * \param[in] context Measurement context
+ * \param[in] interval Trace interval for the reports in milliseconds
+ * \param[in] trace_class Trace class
+ * \param[in] str_prefix_fmt String prefix format to use in the reported trace
+ * \param[in] ... Parameters for the provided prefix format (max. 2 params)
+ */
+#define measure_report(context, interval, trace_class, str_prefix_fmt, ...) \
```

Comment:
You might not need the 'trace_class' once it is moved as part of the scheduler functionality (as suggested in another comment).

Reply:
While that is true it would unnecessarily limit the use cases for this. I'd like the users to be able to measure any part of the code, not just what's handled by a task.